### PR TITLE
 Add syslog tls changes - EV-2481

### PR DESCRIPTION
**Pull request by vara2504**
## Description

This PR has 2 changes.
1) Provide option to enable TLS for syslog forwarding.
2) Update the tigera-ca-bundle fluentd  specific paths (details below).

**Enable TLS option for rsyslog.**
Design Doc : https://docs.google.com/document/d/16xxQEbZXRM-7U52LFvLmm589GIypUyaS/edit#heading=h.gjdgxs
Story: https://tigera.atlassian.net/browse/EV-2481

Testing:
1. Create a Rsyslog server (We can clone a existing [rene micro server](https://docs.google.com/document/d/1XRJoxXYGRVdDQUn28iAwBAezB0VKeg0Qqo2kv_Jyuxg/edit#heading=h.7kalzx4f89n6) using "create similar" option in GCP ). This creates a new micro server with rsyslog installed.
2. Follow Syslog step in [rene's doc](https://docs.google.com/document/d/1XRJoxXYGRVdDQUn28iAwBAezB0VKeg0Qqo2kv_Jyuxg/edit#heading=h.7kalzx4f89n6) to set up the Syslog fowarding using TCP. Update the your box's configuration /etc/rsyslog.conf by mimic the one from rene's box.
3. To Test TLS, apply the below yaml to your cluster
```
apiVersion: operator.tigera.io/v1
kind: LogCollector
metadata:
  name: tigera-secure
spec:
  additionalStores:
    syslog:
      endpoint: tcp://hostname:port  // hostname -I in server box to find the IP , port is anything configured (eg:6514)
      encryption: TLS
      logTypes: 
        - Flows
        - DNS
        - Audit
        - IDSEvents
  collectProcessPath: Enabled
```
4. Install gnutls-bin and rsyslog-gnutls.
     sudo apt-get install gnutls-bin
    sudo apt-get install  rsyslog-gnutls

5. Create self signed CA, key, cert in rsyslog server using certtool. Copy the CA to cluster by creating a secret.
    copy CA into tls.crt file in the cluster and create configmap in operator namespace.
   ``` k create configmap syslog-ca --from-file=tls.crt -n tigera-operator  ```

6. Update the rsyslog server "/etc/rsyslog.conf" with below values.
```
# comment the exiting TCP lines and the below

# provides TCP syslog reception
module(load="imtcp"
        StreamDriver.Name="gtls"
        StreamDriver.Mode="1"
        StreamDriver.Authmode="anon"
)
input(type="imtcp" port="6514")

$DefaultNetstreamDriver gtls
$DefaultNetstreamDriverCAFile /etc/path_to_cert/ca.pem
$DefaultNetstreamDriverCertFile /etc/path_to_cert/server-cert.pem
$DefaultNetstreamDriverKeyFile /etc/path_to_cert/server-key.pem

```
7) Restart and watch for logs.
sudo systemctl restart rsyslog
sudo systemctl status rsyslog
tail -n 1 -f /var/log/syslog

**Update the tigera-ca-bundle fluentd  specific paths.**
Today tigera-ca-bundle is mounted at 2 places

[/etc/fluentd/elastic](https://github.com/tigera/operator/blob/4ba68f436cd2c79311272f72e4a8a41bf7f37845/pkg/render/fluentd.go#L487)
[/etc/pki/tls/certs/](https://github.com/tigera/operator/blob/4ba68f436cd2c79311272f72e4a8a41bf7f37845/pkg/render/fluentd.go#L516)
While mounting on "/etc/pki/tls/certs/" existing Internet ca-bundle.crt which was bundle as part of [tigera fluentd docker image ](https://github.com/tigera/fluentd-docker/blob/master/Dockerfile#L29)is overwritten.

Hence replacing all the tigera-ca-bundle reference to new /etc/fluentd/elastic

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

Need to be merged along with https://github.com/tigera/fluentd-docker/pull/293

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/render/fluentd.go [hunk 1]
```diff
@@ -675,6 +736,16 @@ func (c *fluentdComponent) envvars() []corev1.EnvVar {
 					}
 				}
 			}
+			// Enable TLS option for syslog forwarding.
+			if syslog.Encryption == operatorv1.EncryptionTLS && len(c.cfg.SysLogCredential.Certificate) != 0 {
+				envs = append(envs,
+					corev1.EnvVar{Name: "SYSLOG_TLS", Value: "true"},
+				)
+				envs = append(envs,
+					corev1.EnvVar{Name: "SYSLOG_VERIFY_MODE", Value: "OPENSSL::SSL::VERIFY_NONE"},
```
Comment (matthewdupre) [inline pkg/render/fluentd.go:745]: This should be removed

**Review by matthewdupre — COMMENTED** [review]
Posted some comments - I understand it's a WIP though

File: api/v1/logcollector_types.go [hunk 2]
```diff
@@ -45,6 +45,13 @@ const (
 	CollectProcessPathDisable CollectProcessPathOption = "Disabled"
 )
 
+type EncryptionOption string
+
+const (
+	EncryptionNone EncryptionOption = "none"
```
[the remark this answers is not in the document]
Comment (vara2504) [reply api/v1/logcollector_types.go]: Done

File: api/v1/logcollector_types.go [hunk 3]
```diff
@@ -128,6 +135,13 @@ type SyslogStoreSpec struct {
 	// LogTypes contains a list of types of logs to export to syslog. By default, if this field is
 	// omitted, it will be set to include all possible values.
 	LogTypes []SyslogLogType `json:"logTypes"`
+
+	// Configuration for enabling/disabling TLS option.
+	// If tls, this feature enable tls.
+	// Default: none
```
[the remark this answers is not in the document]
Comment (vara2504) [reply api/v1/logcollector_types.go]: Done

File: pkg/controller/logcollector/logcollector_controller.go [hunk 4]
```diff
@@ -221,7 +221,8 @@ func fillDefaults(instance *operatorv1.LogCollector) []string {
 					v1.SyslogLogDNS,
 					v1.SyslogLogFlows,
 				}
-
+				// Syslog changes
```
[the remark this answers is not in the document]
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go]: Removed the default initialising since expect logTypes no other fields are initialized. Will added it again if suggested.

File: pkg/controller/logcollector/logcollector_controller.go [hunk 5]
```diff
@@ -429,6 +430,23 @@ func (r *ReconcileLogCollector) Reconcile(ctx context.Context, request reconcile
...
+		if instance.Spec.AdditionalStores.Syslog != nil {
+			syslogCredential, err = getSysLogCredential(r.client)
+			if err != nil {
+				log.Error(err, "Error with syslog credential secret")
+				r.status.SetDegraded("Error with syslog credential secret", err.Error())
+				return reconcile.Result{}, err
+			}
+			if syslogCredential == nil {
+				log.Info("syslog credential secret does not exist")
+				r.status.SetDegraded("syslog credential secret does not exist", "")
```
[the remark this answers is not in the document]
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go]: Added the changes for Internet root CA, Hence this error message is no longer applicable.

File: pkg/controller/logcollector/logcollector_controller.go [hunk 6]
```diff
@@ -691,6 +711,54 @@ func getSplunkCredential(client client.Client) (*render.SplunkCredential, error)
 	}, nil
 }
 
+//copies the secret for syslog
+func getSysLogCredential(client client.Client) (*render.SysLogCredential, error) {
+	//tokenSecret := &corev1.Secret{}
```
[the remark this answers is not in the document]
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go]: Done

File: pkg/render/fluentd.go (hunk 1 above)
Comment (vara2504) [reply pkg/render/fluentd.go]: Done

File: pkg/controller/logcollector/logcollector_controller.go [hunk 7]
```diff
@@ -691,6 +711,54 @@ func getSplunkCredential(client client.Client) (*render.SplunkCredential, error)
...
+	certificateSecret := &corev1.Secret{}
+	certificateNamespacedName := types.NamespacedName{
+		Name:      render.SysLogCertificateSecretName,
+		Namespace: common.OperatorNamespace(),
+	}
+
+	if err := client.Get(context.Background(), certificateNamespacedName, certificateSecret); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info(fmt.Sprintf("Syslog certificate secret %v not provided. Assuming http protocol or trusted CA certificate.",
+				render.SysLogCertificateSecretName))
```
[the remark this answers is not in the document]
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go:719]: made changes to return nil,nil for Internet CA bundle scenario

File: pkg/render/fluentd.go (hunk 1 above)
Comment (vara2504) [reply pkg/render/fluentd.go]: Updated the  "SYSLOG_VERIFY_MODE" to Value: "1" ("OPENSSL::SSL::VERIFY_PEER"). This option let the client (fluentd) verify the Cert provided by rsyslog server.

File: api/v1/logcollector_types.go [hunk 8]
```diff
@@ -128,6 +135,13 @@ type SyslogStoreSpec struct {
 	// LogTypes contains a list of types of logs to export to syslog. By default, if this field is
 	// omitted, it will be set to include all possible values.
 	LogTypes []SyslogLogType `json:"logTypes"`
+
+	// Configuration for enabling/disabling TLS option.
```
Comment (rene-dekker) [inline api/v1/logcollector_types.go:139]: How about:

```
// Encryption configures traffic encryption to the Syslog server.
// If TLS, this feature enable tls.  <= maybe remove this line?
```

File: pkg/controller/logcollector/logcollector_controller.go [hunk 9]
```diff
@@ -429,6 +428,18 @@ func (r *ReconcileLogCollector) Reconcile(ctx context.Context, request reconcile
 		}
 	}
 
+	var syslogCredential *render.SysLogCredential
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:431]: I think this is not correct:
- A secret is used, but the contents are a certificate only. This is not secret information. I think the certificate could be part of the LogCollector Spec, much like how you can add certificatePEM for Installation.Spec.CertificateManagement. @tmjd WDYT?
- For this reason `syslogCredential` seems to be the wrong variable name as well.

File: pkg/controller/logcollector/logcollector_controller.go [hunk 10]
```diff
@@ -551,20 +563,21 @@ func (r *ReconcileLogCollector) Reconcile(ctx context.Context, request reconcile
...
-			OSType:          rmeta.OSTypeWindows,
-			TrustedBundle:   trustedBundle,
-			ManagedCluster:  managedCluster,
-			UsePSP:          r.usePSP,
+			LogCollector:     instance,
+			ESSecrets:        esSecrets,
+			ESClusterConfig:  esClusterConfig,
+			S3Credential:     s3Credential,
+			SplkCredential:   splunkCredential,
+			SysLogCredential: syslogCredential,
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:571]: I think we don't need this Syslogcredential at all:
- If the user brings a certificate, we can add the PEM to the TrustedBundle. (TrustedBundle.AddCertificates(...) and certificatemanagement.NewCertificate(...) )
- Then the renderer can simply mount the TrustedBundle in the location that syslog reads from.

This way you can remove a lot of code and continue to follow consistent code practices.

File: api/v1/logcollector_types.go [hunk 11]
```diff
@@ -45,6 +45,13 @@ const (
 	CollectProcessPathDisable CollectProcessPathOption = "Disabled"
 )
 
+type EncryptionOption string
+
+const (
+	EncryptionNone EncryptionOption = "None"
```
Comment (rene-dekker) [inline api/v1/logcollector_types.go:51]: I think we want to set EncryptionNone in the controller if the user omits a value for Encryption. We do this for most other unspecified enums. Usually the method is called fillDefaults(...)

Comment (vara2504) [reply api/v1/logcollector_types.go:51]: Done

File: pkg/controller/logcollector/logcollector_controller.go (hunk 9 above)
Comment (tmjd) [reply pkg/controller/logcollector/logcollector_controller.go]: It looks like this has been discussed in the design document, I think @caseydavenport would be better to consult here since he has been reviewing the design doc.

**Comment by vara2504** [comment]
@rene-dekker Addressed all the review comments.

File: pkg/controller/logcollector/logcollector_controller.go [hunk 12]
```diff
@@ -221,7 +222,9 @@ func fillDefaults(instance *operatorv1.LogCollector) []string {
 					v1.SyslogLogDNS,
 					v1.SyslogLogFlows,
 				}
-
+				if syslog.LogTypes != nil || len(syslog.Encryption) == 0 {
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:225]: Shouldnt this be after of the if-statement of line 216?

And then simply set EncryptionNone if len(syslog.Encryption) == 0 ?

File: pkg/controller/logcollector/logcollector_controller.go [hunk 13]
```diff
@@ -429,6 +432,22 @@ func (r *ReconcileLogCollector) Reconcile(ctx context.Context, request reconcile
 		}
 	}
 
+	var useUserCA bool
+	if instance.Spec.AdditionalStores != nil {
+		if instance.Spec.AdditionalStores.Syslog != nil {
+			syslogCert, err := getSysLogCertificate(r.client)
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:438]: Shouldn't this only happen if encryption == TLS?

File: pkg/controller/logcollector/logcollector_controller.go [hunk 14]
```diff
@@ -753,3 +774,24 @@ func getEksCloudwatchLogConfig(client client.Client, interval int32, region, gro
...
 }
+func getSysLogCertificate(client client.Client) (certificatemanagement.CertificateInterface, error) {
+	cm := &corev1.ConfigMap{}
+	cmNamespacedName := types.NamespacedName{
+		Name:      render.SyslogCAConfigMapName,
+		Namespace: common.OperatorNamespace(),
+	}
+	if err := client.Get(context.Background(), cmNamespacedName, cm); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info(fmt.Sprintf("ConfigMap %q is not found. Assuming internet trusted CA certificate", render.SyslogCAConfigMapName))
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:785]: Maybe rephrase to "ConfigMap %q is not found, assuming Syslog's certificate is signed by a publicly trusted CA"

File: pkg/render/common/elasticsearch/decorator.go [hunk 15]
```diff
@@ -28,6 +28,8 @@ import (
 const (
 	elasticsearchSecretsAnnotation   = "hash.operator.tigera.io/elasticsearch-secrets"
 	elasticsearchConfigMapAnnotation = "hash.operator.tigera.io/elasticsearch-configmap"
+	TigeraCertBundleMountPathWindows = "c:/etc/fluentd/elastic/tigera-ca-bundle.crt"
+	TigeraCertBundleMountPath        = "/etc/fluentd/elastic/tigera-ca-bundle.crt"
```
Comment (rene-dekker) [inline pkg/render/common/elasticsearch/decorator.go:32]: Please revert the changes in this file. this is not the right way to change this.

File: pkg/render/fluentd.go [hunk 16]
```diff
@@ -90,6 +96,10 @@ const (
 
 	PacketCaptureAPIRole        = "packetcapture-api-role"
 	PacketCaptureAPIRoleBinding = "packetcapture-api-role-binding"
+
+	// Verification Mode for syslog forwarding.
```
Comment (rene-dekker) [inline pkg/render/fluentd.go:100]: We can get rid of these values and also the verify_mode, so the default (secure) verification is always used.

File: pkg/render/fluentd.go [hunk 17]
```diff
@@ -157,6 +167,8 @@ type FluentdConfiguration struct {
 
 	// Whether or not the cluster supports pod security policies.
 	UsePSP bool
+	// Whether to use User provided certificate or not.
+	UseUserCA bool
```
Comment (rene-dekker) [inline pkg/render/fluentd.go:171]: Should be called UseUserCertificate, since we do not know if the user supplies a cert or a CA cert

File: pkg/render/fluentd.go [hunk 18]
```diff
@@ -732,7 +760,7 @@ func (c *fluentdComponent) envvars() []corev1.EnvVar {
 
 	if c.SupportedOSType() != rmeta.OSTypeWindows {
 		envs = append(envs,
-			corev1.EnvVar{Name: "CA_CRT_PATH", Value: c.cfg.TrustedBundle.MountPath()},
+			corev1.EnvVar{Name: "CA_CRT_PATH", Value: TigeraCertBundleMountPath},
```
Comment (rene-dekker) [inline pkg/render/fluentd.go:763]: Syslog can be used on windows hosts, so you also need to add the env now for windows.

File: pkg/controller/logcollector/logcollector_controller.go (hunk 10 above)
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go]: Done

File: pkg/render/fluentd.go (hunk 18 above)
Comment (vara2504) [reply pkg/render/fluentd.go:773]: Looks like [this code](https://github.com/tigera/operator/blob/62170a80071ec2bac8822612950c1c5ef55bf1f6/pkg/render/fluentd.go#L733) is intent to populate the EnvVar for !Windows usecase.  Also "[Should render for windows nodes" ](https://github.com/tigera/operator/blob/62170a80071ec2bac8822612950c1c5ef55bf1f6/pkg/render/fluentd_test.go#L193) UT fails if we populate these values for Windows usecase

File: pkg/render/fluentd.go (hunk 17 above)
Comment (vara2504) [reply pkg/render/fluentd.go]: Updated

File: pkg/render/fluentd.go (hunk 16 above)
Comment (vara2504) [reply pkg/render/fluentd.go]: Updated

File: pkg/render/fluentd.go (hunk 18 above)
Comment (vara2504) [reply pkg/render/fluentd.go:773]: Updated

File: pkg/render/common/elasticsearch/decorator.go (hunk 15 above)
Comment (vara2504) [reply pkg/render/common/elasticsearch/decorator.go]: Done

File: pkg/render/fluentd.go (hunk 18 above)
Comment (vara2504) [reply pkg/render/fluentd.go:773]: DOne

File: pkg/controller/logcollector/logcollector_controller.go (hunk 12 above)
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go]: Yes.. updated the code

File: pkg/controller/logcollector/logcollector_controller.go (hunk 14 above)
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go]: Done

**Comment by vara2504** [comment]
@rene-dekker As discussed, updated the tigera-ca-bundle path to /tigera-ca-bundle/tigera-ca-bundle.crt

File: pkg/controller/logcollector/logcollector_controller.go (hunk 13 above)
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go:438]: yes.. Updated

File: pkg/controller/logcollector/logcollector_controller.go [hunk 19]
```diff
@@ -221,10 +222,12 @@ func fillDefaults(instance *operatorv1.LogCollector) []string {
 					v1.SyslogLogDNS,
 					v1.SyslogLogFlows,
 				}
-
 				// Include the field that was modified (in case we need to display error messages)
 				modifiedFields = append(modifiedFields, "AdditionalStores.Syslog.LogTypes")
 			}
+			if len(syslog.Encryption) == 0 {
+				instance.Spec.AdditionalStores.Syslog.Encryption = v1.EncryptionNone
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:229]: I think you want to also add the field to the modifiedFields list, like the others are.

File: pkg/controller/logcollector/logcollector_controller.go [hunk 20]
```diff
@@ -551,20 +571,21 @@ func (r *ReconcileLogCollector) Reconcile(ctx context.Context, request reconcile
...
+			Filters:            filters,
+			EKSConfig:          eksConfig,
+			PullSecrets:        pullSecrets,
+			Installation:       installation,
+			ClusterDomain:      r.clusterDomain,
+			OSType:             rmeta.OSTypeWindows,
+			TrustedBundle:      trustedBundle,
+			ManagedCluster:     managedCluster,
+			UsePSP:             r.usePSP,
+			UseUserCertificate: useUserCertificate,
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:588]: Maybe the variable name could include a hint that this is for syslog. Like `userSyslogCertificate` perhaps?

File: pkg/controller/logcollector/logcollector_controller.go [hunk 21]
```diff
@@ -753,3 +774,24 @@ func getEksCloudwatchLogConfig(client client.Client, interval int32, region, gro
...
 }
+func getSysLogCertificate(client client.Client) (certificatemanagement.CertificateInterface, error) {
+	cm := &corev1.ConfigMap{}
+	cmNamespacedName := types.NamespacedName{
+		Name:      render.SyslogCAConfigMapName,
+		Namespace: common.OperatorNamespace(),
+	}
+	if err := client.Get(context.Background(), cmNamespacedName, cm); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info(fmt.Sprintf("ConfigMap %q is not found,assuming syslog's certificate is signed by publicly trusted CA", render.SyslogCAConfigMapName))
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:785]: nit: space after comma

File: pkg/controller/logcollector/logcollector_controller.go [hunk 22]
```diff
@@ -753,3 +774,24 @@ func getEksCloudwatchLogConfig(client client.Client, interval int32, region, gro
...
+			return nil, nil
+		}
+		return nil, fmt.Errorf("Failed to read ConfigMap %q: %s", render.SyslogCAConfigMapName, err)
+	}
+	if len(cm.Data[render.SyslogCABundleName]) == 0 {
+		log.Info(fmt.Sprintf("ConfigMap %q does not have a field named %q, ,assuming syslog's certificate is signed by publicly trusted CA", render.SyslogCAConfigMapName, render.SyslogCABundleName))
```
Comment (rene-dekker) [inline pkg/controller/logcollector/logcollector_controller.go:791]: nit: two comma's

File: pkg/render/fluentd.go [hunk 23]
```diff
@@ -70,6 +70,11 @@ const (
 	SplunkFluentdSecretsVolName              = "splunk-certificates"
 	SplunkFluentdDefaultCertDir              = "/etc/ssl/splunk/"
 	SplunkFluentdDefaultCertPath             = SplunkFluentdDefaultCertDir + SplunkFluentdSecretCertificateKey
+	SysLogInternetCADir                      = "/etc/pki/tls/certs/"
```
Comment (rene-dekker) [inline pkg/render/fluentd.go:73]: Perhaps "PublicCA" would be a more appropriate name than "InternetCA"

File: pkg/render/fluentd.go [hunk 24]
```diff
@@ -70,6 +70,11 @@ const (
...
+	SysLogInternetCertKey                    = "ca-bundle.crt"
+	SysLogInternetCAPath                     = SysLogInternetCADir + SysLogInternetCertKey
+	SyslogCAConfigMapName                    = "syslog-ca"
+	SyslogCABundleName                       = "tls.crt"
```
Comment (rene-dekker) [inline pkg/render/fluentd.go:77]: We can delete this constant and use the one from the corev1 package: TLSCertKey

File: pkg/controller/logcollector/logcollector_controller.go (hunk 19 above)
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go:229]: Done.

File: pkg/controller/logcollector/logcollector_controller.go (hunk 20 above)
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go:588]: updated

File: pkg/controller/logcollector/logcollector_controller.go (hunk 21 above)
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go:785]: removed the space

File: pkg/controller/logcollector/logcollector_controller.go (hunk 22 above)
Comment (vara2504) [reply pkg/controller/logcollector/logcollector_controller.go:791]: removed

File: pkg/render/fluentd.go (hunk 23 above)
Comment (vara2504) [reply pkg/render/fluentd.go:73]: Done

File: pkg/render/fluentd.go (hunk 24 above)
Comment (vara2504) [reply pkg/render/fluentd.go:77]: Done

File: pkg/render/fluentd_test.go [hunk 25]
```diff
@@ -448,6 +448,144 @@ var _ = Describe("Tigera Secure Fluentd rendering tests", func() {
...
+
+		var volnames []string
+		for _, vol := range ds.Spec.Template.Spec.Volumes {
+			volnames = append(volnames, vol.Name)
+		}
+		Expect(volnames).To(ContainElement("tigera-ca-bundle"))
+
+		envs := ds.Spec.Template.Spec.Containers[0].Env
+
+		expectedEnvs := []struct {
```
Comment (rene-dekker) [inline pkg/render/fluentd_test.go:481]: Wouldn't it be possible to shorten the upcoming 30 lines by using:

```
		Expect(envs).To(ContainElements([]corev1.EnvVar{
			...
		}))
```

**Review by rene-dekker — APPROVED** [review]
lgtm
